### PR TITLE
feat: support use sql to construct graph for loader

### LINKED ISSUE #262: [Feature] Support custom sql for hugegraph-loader module
### Feature Description (功能描述)

Add sql option for mysql which can make the definition of graph data structure more flexible.

### PR DESCRIPTION
**Pull request by simon824**
closed #262

### CODE REVIEW

**Comment by imbajin** [comment]
could u add a user case for it? (in the comment area)

**Comment by simon824** [comment]
> could u add a user case for it? (in the comment area)

1. Vertices/edges and properties are not in the same table, need to `join` them.
2. Need to apply `sql function` on the specified column.

**Comment by imbajin** [comment]
> > could u add a user case for it? (in the comment area)
> 
> 1. Vertices/edges and properties are not in the same table, need to `join` them.
> 2. Need to apply `sql function` on the specified column.

Thanks. Do we have a minimal sql case for it? And it's useful for user-doc later

**Comment by simon824** [comment]
> Thanks. Do we have a minimal sql case for it? And it's useful for user-doc later

```sql
select t1.name, t1.age, t2.city from t1 left join t2 on t1.id = t2.id where t2.date > '2021-01-01'
```
I can help to update doc later.

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java [hunk 1]
```diff
@@ -59,16 +61,24 @@ public void check() throws IllegalArgumentException {
...
+        E.checkArgument(this.table != null || this.sql != null, "At least one of table and sql can't be null");
 
         this.schema = this.vendor.checkSchema(this);
         if (this.driver == null) {
             this.driver = this.vendor.defaultDriver();
         }
     }
 
+    public String sql() {
+        return this.sql;
```
Comment (javeme) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java:75]: can we rename to customSQL or customSelect?

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/reader/jdbc/RowFetcher.java [hunk 2]
```diff
@@ -119,13 +130,17 @@ public List<Line> nextBatch() throws SQLException {
...
+        String select = this.source.existsSql() ?
+                this.source.sql() : this.source.vendor().buildSelectSql(this.source, this.nextStartRow);
+
         LOG.debug("The sql for select is: {}", select);
 
         List<Line> batch = new ArrayList<>(this.source.batchSize() + 1);
         try (Statement stmt = this.conn.createStatement();
              ResultSet result = stmt.executeQuery(select)) {
+            if (this.source.existsSql()) {
+                readHeader(result);
```
Comment (javeme) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/reader/jdbc/RowFetcher.java:142]: prefer to keep style: "this.readHeader(result);"

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/reader/jdbc/RowFetcher.java [hunk 3]
```diff
@@ -95,6 +96,16 @@ public String[] readHeader() throws SQLException {
         return this.columns;
     }
 
+    public String[] readHeader(ResultSet rs) throws SQLException {
```
Comment (javeme) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/reader/jdbc/RowFetcher.java:99]: can mark as private?

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java [hunk 4]
```diff
@@ -63,20 +63,20 @@ public void check() throws IllegalArgumentException {
         E.checkArgument(this.database != null, "The database can't be null");
         E.checkArgument(this.username != null, "The username can't be null");
         E.checkArgument(this.password != null, "The password can't be null");
-        E.checkArgument(this.table != null || this.sql != null, "At least one of table and sql can't be null");
+        E.checkArgument(this.table != null || this.customSQL != null, "At least one of table and sql can't be null");
```
Comment (javeme) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java:66]: wrap line after "," to keep 80 chars each line.

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java [hunk 5]
```diff
@@ -63,20 +63,20 @@ public void check() throws IllegalArgumentException {
...
 
-    public String sql() {
-        return this.sql;
+    public String customSQL() {
+        return this.customSQL;
     }
 
     public boolean existsSql() {
-        return this.sql != null;
+        return this.customSQL != null;
```
Comment (javeme) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java:79]: could you please also update existsSql() name?

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/reader/jdbc/RowFetcher.java [hunk 6]
```diff
@@ -131,15 +131,15 @@ public List<Line> nextBatch() throws SQLException {
         }
 
         String select = this.source.existsSql() ?
-                this.source.sql() : this.source.vendor().buildSelectSql(this.source, this.nextStartRow);
+                this.source.customSQL() : this.source.vendor().buildSelectSql(this.source, this.nextStartRow);
```
Comment (javeme) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/reader/jdbc/RowFetcher.java:134]: can we adjust code style: https://github.com/apache/incubator-hugegraph-doc/wiki/HugeGraph%E4%BB%A3%E7%A0%81%E9%A3%8E%E6%A0%BC%E6%8C%87%E5%8D%97

Comment (simon824) [reply hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/reader/jdbc/RowFetcher.java]: get it ,thx.

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java [hunk 7]
```diff
@@ -75,7 +76,7 @@ public String customSQL() {
         return this.customSQL;
     }
 
-    public boolean existsSql() {
+    public boolean existsSQL() {
```
Comment (javeme) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java:79]: prefer customizedSQL existsCustomSQL

**Review by javeme — DISMISSED** [review]

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java (hunk 7 above)
Comment (simon824) [reply hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/source/jdbc/JDBCSource.java]: Updated, thx

**Review by javeme — APPROVED** [review]

**Review by imbajin — APPROVED** [review]
Thanks, u could add a doc later

**Pull request merged**